File: Src/Icm.ContextConsole/Context/RootContext.vb
Imports Icm.Tree

Imports Icm.Localization

''' <summary>
''' This base root context provides implementation of Help action and Quit action.
''' </summary>
''' <remarks></remarks>
Public MustInherit Class RootContext
    Inherits BaseContext

    Private Application As IApplication

    Protected Overrides Sub BaseInitialize(app As IApplication)
        MyBase.BaseInitialize(app)
        Application = app
    End Sub

    Public MustOverride Sub Credits()

    Public Sub Help()
        ' Actions of current context
        ShowActions(Application.CurrentContextNode.Value)

        ' Available subcontexts
        Interactor.ShowList(
            "Subcontexts",
            Application.CurrentContextNode.GetChildNodes,
            Function(ctx)
                Return String.Format(My.Resources.Resources.root_use, _
                        ctx.Value.Name.ToLower)
            End Function,
            hideIfEmpty:=True)

        ' Inherited actions from ancestors
        For Each ctx In Application.CurrentContextNode.ProperAncestors
            ShowActions(ctx)
        Next
    End Sub

    Public Sub Contexts()
        For Each contextAndLevel In Application.RootContextNode.DepthPreorderTraverseWithLevel
            Interactor.ShowMessage(New String(" "c, contextAndLevel.Level) & contextAndLevel.Result.Name)
        Next
    End Sub

    Public Overridable Sub Quit()
        Interactor.ShowMessage(locRepo.Trans("quit_bye"))
    End Sub

    Private Sub ShowActions(ByVal ctl As IContext)
        Dim IsCurrent = (Application.CurrentContextNode.Value.Name = ctl.Name)

        Dim title As String
        If IsCurrent Then
            title = locRepo.TransF("help_title", ctl.Name, PhrF("help_current"))
        Else
            title = locRepo.TransF("help_title", ctl.Name)
        End If

        Interactor.ShowList(
            title,
            ctl.GetActions(),
            Function(action) _
                     String.Format("{0}: {1}", _
                        action.Name.ToLower, _
                        If( _
                            TranslateActionDescription(action), _
                            action.Name)))
    End Sub

    Private Function TranslateActionDescription(action As IAction) As String
        If action.IsInternal Then
            Return locRepo.Trans(action.LocalizationKey)
        Else
            Return extLocRepo.Trans(action.LocalizationKey)
        End If
    End Function

End Class

File: Src/Icm.ContextConsole/Context/RootContext.cs
using Icm.Tree;

using Icm.Localization;
using static Icm.Localization.PhraseFactory;

/// <summary>
/// This base root context provides implementation of Help action and Quit action.
/// </summary>
/// <remarks></remarks>
public abstract class RootContext : BaseContext
{


	private IApplication _application;
	protected override void BaseInitialize(IApplication app)
	{
		base.BaseInitialize(app);
		_application = app;
	}

	public abstract void Credits();

	public void Help()
	{
		// Actions of current context
		ShowActions(_application.CurrentContextNode.Value);

		// Available subcontexts
		Interactor.ShowList("Subcontexts", 
            _application.CurrentContextNode.GetChildNodes(), 
            ctx => string.Format(My.Resources.Resources.root_use, ctx.Value.Name().ToLower()), 
            hideIfEmpty: true);

		// Inherited actions from ancestors
		foreach (var ctx in _application.CurrentContextNode.ProperAncestors()) {
			ShowActions(ctx);
		}
	}

	public void Contexts()
	{
		foreach (var contextAndLevel in _application.RootContextNode.DepthPreorderTraverseWithLevel()) {
			Interactor.ShowMessage(new string(' ', contextAndLevel.Level) + contextAndLevel.Result.Name());
		}
	}

	public virtual void Quit()
	{
		Interactor.ShowMessage(locRepo.Trans("quit_bye"));
	}

	private void ShowActions(IContext ctl)
	{
		bool isCurrent = _application.CurrentContextNode.Value.Name() == ctl.Name();

		string title;
		title = isCurrent 
            ? locRepo.TransF("help_title", ctl.Name(), PhrF("help_current")) 
            : locRepo.TransF("help_title", ctl.Name());

		Interactor.ShowList(title, ctl.GetActions(), action =>
		    $"{action.Name().ToLower()}: {TranslateActionDescription(action) ?? action.Name()}");
	}

	private string TranslateActionDescription(IAction action)
	{
		if (action.IsInternal()) {
			return locRepo.Trans(action.LocalizationKey());
		} else {
			return extLocRepo.Trans(action.LocalizationKey());
		}
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
